src/snowflake/snowpark/mock/functions.py

```diff
@@ -141,22 +143,34 @@ def mock_sum(column: ColumnEmulator) -> ColumnEmulator:
 
 @patch("avg")
 def mock_avg(column: ColumnEmulator) -> ColumnEmulator:
-    all_item_is_none = True
-    ret = 0
-    cnt = 0
-    for data in column:
-        if data is not None:
-            all_item_is_none = False
-            ret += float(data)
-            cnt += 1
+    if not isinstance(column.sf_type.datatype, (_NumericType, NullType)):
+        raise SnowparkSQLException(
+            f"Cannot compute avg on a column of type {column.sf_type.datatype}"
+        )
 
-    ret = (
-        ColumnEmulator(data=[round((ret / cnt), 5)])
-        if not all_item_is_none
-        else ColumnEmulator(data=[None])
-    )
-    ret.sf_type = column.sf_type
-    return ret
+    if isinstance(column.sf_type.datatype, NullType) or column.isna().all():
+        return ColumnEmulator(data=[None], sf_type=ColumnType(NullType(), True))
+    elif isinstance(column.sf_type.datatype, _IntegralType):
+        res_type = DecimalType(38, 6)
+    elif isinstance(column.sf_type.datatype, DecimalType):
+        precision, scale = (
+            column.sf_type.datatype.precision,
+            column.sf_type.datatype.scale,
+        )
+        precision = max(38, column.sf_type.datatype.precision + 12)
+        if scale <= 6:
+            scale = scale + 6
+        elif scale < 12:
+            scale = 12
+        res_type = DecimalType(precision, scale)
+    else:
```

Comment:
How about an explicit checking for `FloatType` and `DoubleType`?

Reply:
That's a good idea. Do you prefer using
```
elif instance(column.sf_type.datatype, _FractionalType)
    ...
else:
   raise Error
```
or an assert statement in the else case on line 166?